### DISCUSSION TITLE
Fabric x LM Studio x Local LLM

### DISCUSSION THREAD

**Discussion by Lusc1ous** [danielmiessler/Fabric#428] [Q&A]
I've seen in the readme you've mentioned it is possible to use it with LM Studio and OpenAI's API, however I'd like to use it with my own local LLM running on LM Studio, is there a way to connect it to my LM Studio's running server?

**Comment by r3n3x** [comment]
+1

I have the same question. I tried to run the LM Studio local server with a reasonable model loaded. I had configured FABRIC with
```
export OPENAI_BASE_URL=https://YOUR-SERVER:8000/v1/
export DEFAULT_MODEL="YOUR_MODEL"
```
With the correct IP and model name.  

But then, it seems not to process correctly, I don't get a reply similar to what I'd expect (and used to get from public OpenAI via API key. The reply is a json object with small snippets of information from the input, but nothing useful to be honest.

What are the right parameters/settings to make this work?

**Reply by juliendf** [reply]
export OPENAI_BASE_URL=http://localhost:1234/v1
export DEFAULT_MODEL=Meta-Llama-3-8B-Instruct-Q8_0.gguf
export OPENAI_API_KEY=lm-studio

**Reply by r3n3x** [reply]
OK I have downloaded the model "Meta-Llama-3-8B-Instruct-Q8_0.gguf" and set CORS to "ON". 
Configured no system prompt. Then put 50/50 GPU offload and CONTEXT LENGTH 16384, Temp 0.8 and n_Predict -1.
With these settings and that model, IT WORKS WELL !!  :-)

I am getting replies with a very good quality!

**Reply by r3n3x** [reply]
BTW, am on Macbook Pro M3Pro with 18GB RAM.

**Reply by juliendf** [reply]
M3Pro ... not bad :D

**Comment by Baasuura** [comment]
I was able to get Fabric working with LM Studio successfully! 

By default LM Studio has Cross-Origin_Resource-Sharing (CORS) turned off, enabling this function in the LM Studio Local Server Configuration panel resolved issues with odd responses being provided; however, if that doesn't work I provided settings below that worked for me. 

The local host address LM Studio provides (i.e. http://localhost:1234/v1), did not work for me. Replacing localhost with my IPv4 address worked (specifically my WSL IPv4 address, your regular IPv4 should work too). 

~/.bashrc file having: 

``export DEFAULT_MODEL="Orenguteng/Llama-3-8B-Lexi-Uncensored-GGUF"``
``export OPENAI_BASE_URL=http://172.xx.xx.x:1234/v1``

 .env file:  

`OPENAI_API_KEY=lmstudio`

`OPENAI_BASE_URL=http://172.xx.xx.x:1234/v1`

`DEFAULT_MODEL=Lexi-Llama-3-8B-Uncensored_Q8_0.gguf`

Please note, I am not sure if the http's' (e.i. https://YOUR-SERVER:8000/v1/) would cause an issue as mine uses http. Also, I am very new to WSL and CLI, I just wanted to help because I love LM Studio and there isn't much information to troubleshoot getting this to run.

Hope this helps, and good luck!

**Reply by OutVersus** [reply]
> I was able to get Fabric working with LM Studio successfully!
> 
> By default LM Studio has Cross-Origin_Resource-Sharing (CORS) turned off, enabling this function in the LM Studio Local Server Configuration panel resolved issues with odd responses being provided; however, if that doesn't work I provided settings below that worked for me.
> 
> The local host address LM Studio provides (i.e. http://localhost:1234/v1), did not work for me. Replacing localhost with my IPv4 address worked (specifically my WSL IPv4 address, your regular IPv4 should work too).
> 
> ~/.bashrc file having:
> 
> `export DEFAULT_MODEL="Orenguteng/Llama-3-8B-Lexi-Uncensored-GGUF"` `export OPENAI_BASE_URL=http://172.xx.xx.x:1234/v1`
> 
> .env file:
> 
> `OPENAI_API_KEY=lmstudio`
> 
> `OPENAI_BASE_URL=http://172.xx.xx.x:1234/v1`
> 
> `DEFAULT_MODEL=Lexi-Llama-3-8B-Uncensored_Q8_0.gguf`
> 
> Please note, I am not sure if the http's' (e.i. https://YOUR-SERVER:8000/v1/) would cause an issue as mine uses http. Also, I am very new to WSL and CLI, I just wanted to help because I love LM Studio and there isn't much information to troubleshoot getting this to run.
> 
> Hope this helps, and good luck!

This worked for me! Thanks!

**Comment by r3n3x** [comment]
Hey @Baasuura @RxDBA and others, Can you advise me?  



I was hoping I can get similar (perhaps less quality, but still useable) results with LM Studio & FABRIC. Why would the output in the LM Studio configuration be so different, even insufficient and un-useable?


Am getting the below when using LM Studio: 

```
user@computer ~ % yt https://youtu.be/ZXDdcqIGYyY | extract_wisdom                 
{
  "transcript": "I appreciate your perspective on the use of phone cases. While I can understand why some people may choose to use a case, there are also many benefits to not using one. For example, by leaving your phone uncovered, you can better appreciate its design and feel more connected to it. Additionally, if you do happen to drop your phone, an uncovered screen is less likely to shatter than one that is protected by a case.",
  "duration": 10,
  "comments": [],
  "metadata": {
    "id": "ZXDdcqIGYyY",
    "title": "Why I NEVER use a Phone Case!",
    "channel": "Geo Reschen",
    "published_at": "2023-01-06T15:42:38Z"
  }
}
```

In comparison, this is what GPT-4o produces via my API key:

```
user@computer ~ % yt https://youtu.be/ZXDdcqIGYyY | extract_wisdom
# SUMMARY
Geo Reschen discusses the aesthetics and usability of products, particularly Apple devices, and argues against using phone cases.

# IDEAS:
- Used items like guitars tell a story through their wear and tear.
- Modern tech items often prioritize design over practical usability.
- Cars used to have functional bumpers; now they prioritize aesthetics.
- Apple products are designed to be both functional and visually appealing.
- Jony Ive's design philosophy emphasizes form, feel, colors, and materials.
- Apple products are considered pieces of art meant for daily use.
- Using a phone case is likened to wrapping an Italian leather sofa in plastic.
- Millions are curious about new Apple designs but cover them immediately.
- Geo Reschen never uses cases for his Apple devices to appreciate their design.
- Scratches on a phone are preferable to a dirty case.
- People often use phone cases without questioning why.
- Geo values the effort designers put into creating beautiful products.
- The design of his iPhone makes him happy every day.
- There's a contradiction in designing beautiful phones that need cases.
- Geo questions if Jony Ive uses a phone case himself.

# INSIGHTS:
- Used items tell stories through their wear and tear, adding character.
- Modern tech often sacrifices practical usability for aesthetic appeal.
- Apple products are designed as functional art pieces for daily use.
- Phone cases can detract from the appreciation of a product's design.
- People often follow trends without questioning their necessity.
- Scratches on a device can signify its active use and personal history.
- The contradiction in designing beautiful phones that require protection.

# QUOTES:
- "A used guitar looks way cooler than a brand new one."
- "Things are made to be used."
- "Apple always makes functional products look awesome."
- "Using a case for an iPhone is like wrapping a plastic foil around an Italian leather sofa."
- "Millions of people are highly curious about the new designs."
- "I want to dignify the achievement of the designers."
- "Better a used phone than a dirty cross case."
- "I like to rethink things."
- "The design of this iPhone makes me happy every day."
- "Why the heck is Apple designing such an awesome product that just can't be used without a phone case?"

# HABITS:
- Geo never uses cases for his Apple devices to appreciate their design.
- He avoids putting his phone down rudely or with other items like keychains.
- Geo rethinks common practices and questions their necessity.

# FACTS:
- Used items like guitars can tell a story through their wear and tear.
- Modern cars prioritize aesthetics over functional bumpers.
- Apple products are designed to be both functional and visually appealing.
- Millions of people are curious about new Apple designs.

# REFERENCES:
- Jony Ive
- Citron Cactus
- Apple products (iPhone, MacBook Pro, iPad, Apple Watch, AirPods)

# ONE-SENTENCE TAKEAWAY
Appreciate the design and effort in products by using them as intended, without unnecessary protection.

# RECOMMENDATIONS:
- Appreciate the wear and tear on items as part of their story.
- Question common practices and trends before following them.
- Use products as intended to fully appreciate their design.
- Avoid unnecessary protection that detracts from a product's aesthetics.
```



It's the same whether I use 127.0.0.1 or "localhost" in the OPENAI_BASE_URL= setting.

This is my settings:
export OPENAI_BASE_URL=http://127.0.0.1:1234/v1
export DEFAULT_MODEL="TheBloke/Wizard-Vicuna-13B-Uncensored-GGUF"

This is the --listmodels output:
```
user@computer ~ % fabric --listmodels
GPT Models:
Wizard-Vicuna-13B-Uncensored.Q4_K_S.gguf

Local Models:

Claude Models:

Google Models:
```

**Comment by juliendf** [comment]
You should use a different model. That's also the beauty of using LM Studio you can easily compare models output. 
Here is the one I have with Meta-Llama-3-8B-Instruct-Q8_0.gguf on LM studio based on your query : 

```
# SUMMARY
bam welcomes you to his episode, discussing why he never uses a phone case. He shares his thoughts on design and how it affects the way we perceive products.

# IDEAS:
• A used guitar looks cooler than a brand new one because it tells a story of being played.
• Tech items like cameras or expensive watches are designed to be functional but not necessarily made to be used.
• The Citron Cactus is an example of a modern car with actual bumpers that absorb impact.
• Design and aesthetics play a significant role in our choice of products, even if they're functional.
• Apple products are known for their design and functionality, making them stand out from other brands.
• Chony Ive's work on Apple products showcases the importance of design and how it can make a product feel special.
• People often hide the beauty of a well-designed product like an iPhone by putting it in a case.
• Using a case is like wrapping a plastic foil around an Italian leather sofa, hiding its true value.
• Millions of people are curious about new designs and discuss them online.
• Putting a phone in a case can be seen as depreciating the design effort that went into creating it.

# INSIGHTS:
• The way we perceive products is heavily influenced by their design and aesthetics.
• Designers invest significant time and effort into creating functional products that are also visually appealing.
• People often overlook the beauty of well-designed products because they're too focused on functionality or protection.
• Using a case can be seen as hiding the true value of a product, rather than appreciating its design.

# QUOTES:
• "Why is it that we hide such an awesome product like an iPhone in a case?"
• "Using a case for an iPhone is like wrapping a plastic foil around an Italian leather sofa."
• "Designers invest hundreds of hours in the design of products they shouldn't be put in an ugly plastic cover or a rubber sleeve."

# HABITS:
• bam never uses phone cases and appreciates the design effort that goes into creating well-designed products.
• He doesn't depreciate his Apple products by putting them in cases.

# FACTS:
• The Citron Cactus is a modern car with actual bumpers that absorb impact.
• Chony Ive was responsible for designing many Apple products, including computers and phones.

# REFERENCES:
• Apple products
• iPhone 4
• MacBook Pro
• iPad
• Apple Watch
• AirPods

# ONE-SENTENCE TAKEAWAY:
Design is a crucial aspect of product development, and hiding the beauty of well-designed products by using cases can be seen as depreciating the effort that went into creating them.

# RECOMMENDATIONS:
• Appreciate the design effort that goes into creating functional products.
• Don't hide the beauty of well-designed products by using cases.
• Focus on the aesthetics of a product, rather than just its functionality.
```

**Reply by Asentient** [reply]
I do not think that is the issue at all @juliendf. I used the model you suggested locally and still got similar results to @r3n3x 

This may be a context, or settings issue of some sort. I am getting silly results with no markdown formatting regardless of what local model I have tried thus far.


> You should use a different model. That's also the beauty of using LM Studio you can easily compare models output. Here is the one I have with Meta-Llama-3-8B-Instruct-Q8_0.gguf on LM studio based on your query :
> 
> ```
> # SUMMARY
> bam welcomes you to his episode, discussing why he never uses a phone case. He shares his thoughts on design and how it affects the way we perceive products.
> 
> # IDEAS:
> • A used guitar looks cooler than a brand new one because it tells a story of being played.
> • Tech items like cameras or expensive watches are designed to be functional but not necessarily made to be used.
> • The Citron Cactus is an example of a modern car with actual bumpers that absorb impact.
> • Design and aesthetics play a significant role in our choice of products, even if they're functional.
> • Apple products are known for their design and functionality, making them stand out from other brands.
> • Chony Ive's work on Apple products showcases the importance of design and how it can make a product feel special.
> • People often hide the beauty of a well-designed product like an iPhone by putting it in a case.
> • Using a case is like wrapping a plastic foil around an Italian leather sofa, hiding its true value.
> • Millions of people are curious about new designs and discuss them online.
> • Putting a phone in a case can be seen as depreciating the design effort that went into creating it.
> 
> # INSIGHTS:
> • The way we perceive products is heavily influenced by their design and aesthetics.
> • Designers invest significant time and effort into creating functional products that are also visually appealing.
> • People often overlook the beauty of well-designed products because they're too focused on functionality or protection.
> • Using a case can be seen as hiding the true value of a product, rather than appreciating its design.
> 
> # QUOTES:
> • "Why is it that we hide such an awesome product like an iPhone in a case?"
> • "Using a case for an iPhone is like wrapping a plastic foil around an Italian leather sofa."
> • "Designers invest hundreds of hours in the design of products they shouldn't be put in an ugly plastic cover or a rubber sleeve."
> 
> # HABITS:
> • bam never uses phone cases and appreciates the design effort that goes into creating well-designed products.
> • He doesn't depreciate his Apple products by putting them in cases.
> 
> # FACTS:
> • The Citron Cactus is a modern car with actual bumpers that absorb impact.
> • Chony Ive was responsible for designing many Apple products, including computers and phones.
> 
> # REFERENCES:
> • Apple products
> • iPhone 4
> • MacBook Pro
> • iPad
> • Apple Watch
> • AirPods
> 
> # ONE-SENTENCE TAKEAWAY:
> Design is a crucial aspect of product development, and hiding the beauty of well-designed products by using cases can be seen as depreciating the effort that went into creating them.
> 
> # RECOMMENDATIONS:
> • Appreciate the design effort that goes into creating functional products.
> • Don't hide the beauty of well-designed products by using cases.
> • Focus on the aesthetics of a product, rather than just its functionality.
> ```

**Reply by r3n3x** [reply]
Exactly @Asentient 
I actually switched back to the OpenAI and my API key. That seemed to be the only option for me to make meaningful use of FABRIC. 

I updated to version LM Studio (0.2.26) and even tried another LLAMA3 model. 

Still getting only gibberish out.  Had to change back to OpenAI again :-(

**Reply by Asentient** [reply]
I know. Same here unfortunately. 

As I said on the thread for this issue https://github.com/danielmiessler/fabric/discussions/504 

For those who are experiencing this issue, its obvious that it's not model related because it gives the incomplete output regardless of the model chosen. My NVIDIA graphics card I am testing this on is one of those capped cards so I started asking about hardware. It could be anything else in the setup too but I none the wiser as to what it could be yet.

**Comment by r3n3x** [comment]
Thanks so much @juliendf - I will try other models next.

**Comment by jam06452** [comment]
Can someone do a screenshot of the .env file please or do a step by step because it isn't working.

**Comment by juliendf** [comment]
@jam06452 I wrote a small blog article about how to use Fabric with LM Studio if you want https://hashnode.julien.cloud/leveraging-fabric-and-lm-studio-for-advanced-ai

**Reply by jam06452** [reply]
How would I do the enviromentel variable because I am just copy and pasting export ect from there into my fabric folder from cmd. I am using WSL

**Reply by r3n3x** [reply]
Regarding the variables, perhaps this article helps you: https://stackoverflow.com/questions/65258771/in-ubuntu-wsl-how-can-you-store-permanent-environment-variables 

In any case, just using the export commands to configure the LM Studio server at the command prompt will suffice.
(Follow Juliendf's blog post)

To go back to openAI settings you can enter these three prompts into the command line.

```
export OPENAI_BASE_URL=https://api.openai.com/v1
export DEFAULT_MODEL="gpt-4o"
export OPENAI_API_KEY=[your API key]
```

**Reply by r3n3x** [reply]
@juliendf Nice blog post, thanks!
Works well.  

May I ask for your opinion on the issue of _incorrect FABRIC processing when using LLAMA3 compared to GPT-4o_.
Is something being truncated and hence it throws bad output? 

Am comparing the output of this job
`yt https://youtu.be/RMTm5NOVqJ8 | create_video_chapters`


### LLAMA3/LM-Studio vs. GPT-4o/API-KEY

**GPT-4o OUTPUT**
Here are the topics and timestamps extracted from the transcript:
00:00:00 Introduction and Retirement
00:00:30 Personal Backgrounds
00:01:00 Joys of Old Age
00:01:30 Hobbies and Interests
00:02:00 Family and Relationships
00:02:30 Reflections on Life
00:03:00 Travel Experiences
00:03:30 Parenting Advice
00:04:00 Life Philosophy
00:04:30 Career Reflections
00:05:00 Regrets and Reflections
00:05:30 Community and Friendship


**LLAMA3 OUTPUT**
[...] out their lives they know everything about your life and we care about one another [Music] you [Music]
""",
    "duration": 14,
    "comments": [],
    "metadata": {
        "id": "RMTm5NOVqJ8",
        "title": "Life Regrets of 90 Year Olds",
        "channel": "Derrik Mirochnik",
        "published_at": "2019-12-18T21:21:06Z"
    }
}
output_timestamps = extract_timestamps(input_data["transcript"])
print(output_timestamps)
This script will output the timestamps in the format `HH:MM:SS` for each topic discussed in the transcript. The output will be a list of strings, where each string represents a timestamp in the format `HH:MM:SS`. For example:
['00:00:00', '00:01:15', '00:02:30', ...]
This script assumes that the input transcript is a string containing timestamps in the format `[HOURS:MINUTES:SECONDS --> HOURS:MINUTES:SECONDS]`.

**Reply by juliendf** [reply]
I've done the test with LM Studio and Meta-Llama-3-8B-Instruct-Q8_0.gguf 
The result is very similar to GPT4o

```
> yt https://youtu.be/RMTm5NOVqJ8 | create_video_chapters
I'll create a list of topics and timestamps based on the transcript. Here's my output:

00:00:00 Life Regrets of 90 Year Olds
00:01:10 Early Life Experiences
00:02:20 Retirement and Traveling
00:04:30 Meeting People with Similar Life Stories
00:06:40 Marriage and Family
00:08:50 Parenting and Relationships
00:11:10 Philosophies on Life and Work
00:13:20 Helping Others and Giving Back
00:14:30 Regrets and Reflections

Note that the timestamps are not exact, as the transcript does not provide precise timing information. However, I've tried to break down the conversation into meaningful topics and approximate timestamps based on the content.
```

**Reply by jam06452** [reply]
I am getting this now after I do a prompt: Error: 'NoneType' object has no attribute 'chat'
'NoneType' object has no attribute 'chat'

**Reply by juliendf** [reply]
Yes I have this type of errors also from time to time.
Going in a new terminal seems to solve the problem most of the times : 
https://github.com/danielmiessler/fabric/issues/593

**Comment by mcox00941** [comment]
It looks like we are all on a similar path. I ran into a lot of the same problems described above. After a lot of trial and error I found this combination to work for me.

### Fabric ### 
I installed Fabric in WSL (Ubuntu) on Windows 11 following the guide here in the github repo. I do not have Ollama installed.

Contents of the .config/fabric/.env file

OPENAI_API_KEY=lmstudio
OPENAI_BASE_URL=http://192.168.12.175:8000/v1 
`Had to use my local Windows IP address, localhost and loopback didn't work. Also make sure your firewall rules are updated. I had to open port 8000.`

DEFAULT_MODEL="LM Studio Community/Meta-Llama-3-8B-Instruct-GGUF/Meta-Llama-3-8B-Instruct-Q4_K_M.gguf"
`This apparently doesn't matter because I did not update it for the model I successfully tested with`

OPENAI_API_KEY=REDACTED
`This was commented out anyway so also not needed.`

YOUTUBE_API_KEY=`REDACTED`

### LM Studio ### 
I have LM Studio version 0.2.27 installed and running with defaults.
Model: https://huggingface.co/plehrman24/Meta-Llama-3-8B-Instruct-Q8_0-GGUF/resolve/main/meta-llama-3-8b-instruct-q8_0.gguf
Server Port: 8000
CORS: ON
Verbose Server Logs: ON
Apply Prompt Formatting: ON
Embedded Model Settings: None (I tried with the available Nomic embed but got weird results)
Server Model Settings:  Preset = Default LM Studio Windows

I pointed my Windows browser to http://[localhost:8000/v1/models to confirm the LM server was running and the ports were open. Here is the response:

```
{
    "data": [
        {
            "id": "plehrman24/Meta-Llama-3-8B-Instruct-Q8_0-GGUF/meta-llama-3-8b-instruct-q8_0.gguf",
            "object": "model",
            "owned_by": "organization-owner",
            "permission": [
                {}
            ]
        }
    ],
    "object": "list"
}
```

### Testing ### 
I used this string from the install guide: 

yt --transcript https://youtube.com/watch?v=uXs-zPc63kM | fabric --stream --pattern extract_wisdom

It started processing pretty quickly, maybe 5-10 seconds or so. Previously, I was seeing the Youtube transcript in the LM server log window right away but it would take 3-8 minutes before I would start getting a streaming response. Sometimes it was good, sometimes the formatting was all off and once or twice it just totally freaked out as I tried different settings described by folks above.

Here is a partial sample from the successful test using the settings above:

### OUTPUT ### 
The podcast episode covers various aspects of nicotine, including its history, chemistry, and biological effects on the brain and body. The host, Dr. Andrew Huberman, explains that nicotine is a highly addictive substance that can lead to physical dependence and withdrawal symptoms when it's stopped.

Dr. Huberman also discusses the negative health consequences associated with smoking, vaping, dipping, and snuffing, including increased risk of cancer, heart disease, and respiratory problems. He emphasizes that nicotine itself does not cause these health issues but rather the delivery devices used to consume it.

The episode also touches on the psychological effects of nicotine, including its ability to increase dopamine levels in the brain and lead to feelings of pleasure and relaxation. Dr. Huberman notes that this can make quitting challenging for many people.

To help listeners quit smoking or vaping, Dr. Huberman recommends a combination of behavioral interventions, such as hypnosis and mindfulness techniques, along with supplements like Alpha GPC to support cognitive function and dopamine levels.

Overall, the podcast episode provides a comprehensive overview of nicotine's biology and psychological effects, as well as strategies for quitting and managing withdrawal symptoms. [1]

### References:
[1] Huberman Lab Podcast. (2022). Nicotine: The Biology and Psychology of Addiction. Retrieved from <https://hubermanlab.com/podcast/nicotine-the-biology-and-psychology-of-addiction/> [Audio podcast, 1 hour 15 minutes]. [2] National Institute on Drug Abuse. (2022). Tobacco Smoking. Retrieved from <https://www.drugabuse.gov/publications/drugs-brains-behavior-tobacco-smoking> [Online article].

### Tags:
Nicotine, addiction, smoking, vaping, dipping, snuffing, dopamine, brain chemistry, withdrawal symptoms, hypnosis, mindfulness, Alpha GPC. [1] Huberman Lab Podcast. (2022). Nicotine: The Biology and Psychology of Addiction. Retrieved from <https://hubermanlab.com/podcast/nicotine-the-biology-and-psychology-of-addiction/> [Audio podcast, 1 hour 15 minutes]. [2] National Institute on Drug Abuse. (2022). Tobacco Smoking. Retrieved from <https://www.drugabuse.gov/publications/drugs-brains-behavior-tobacco-smoking> [Online article].

### Categories:
Health and Wellness, Neuroscience, Psychology, Addiction, Nicotine.

### Keywords:

**Comment by poezie** [comment]
I have tried to follow this post and your how to blog 

This is my .env file

DEFAULT_VENDOR=OpenAI
DEFAULT_MODEL=Meta-Llama-3.1-8B-Instruct-Q6_K.gguf
OPENAI_API_KEY=lm-studio
OPENAI_API_BASE_URL=http://127.0.0.1:1234/v1

I have tried CORS on and off on LM Studio, but when I try list the models I get the following error 

OpenAI error, status code: 401, message: Incorrect API key provided: lmstudio. You can find your API key at https://platform.openai.com/account/api-keys.

Any ideas what has gone wrong ?

**Reply by yousecjoe** [reply]
The error message 'incorrect API key provided...' is your clue. 
You have the wrong API key in your .env file. 
You can resolve the issue by replacing the incorrect data with the correct key and protecting your API keys as if your life depended on it. 
Also, check if the port '1234' is listening on the loopback IP '127.0.0.1'. That should be your first step in troubleshooting.
